### PR TITLE
IGNITE-19641 Catalog events are triggered too early.

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-19641

### CODE REVIEW

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java [hunk 1]
```diff
@@ -1248,19 +1275,22 @@ public void testCreateIndexEvents() {
...
 
         // Drop table.
         assertThat(service.dropTable(dropTableparams), willCompleteSuccessfully());
 
         // Try drop index once again.
         assertThat(service.dropIndex(dropIndexParams), willThrow(IndexNotFoundException.class));
 
-        verifyNoMoreInteractions(eventListener);
+        // Should have destroyed the pk index.
+        verify(eventListener).notify(any(DropIndexEventParameters.class), isNull());
```
Comment (AMashenkov) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java:1293]: It is not clear, if `DropIndexEventParameters` was caused by `dropTable` or `dropIndex` call.
Let's move this higher.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 2]
```diff
@@ -277,82 +259,62 @@ private Catalog catalogAt(long timestamp) {
...
-            params.columns().stream().map(ColumnParams::name).filter(Predicate.not(new HashSet<>()::add))
-                    .findAny().ifPresent(columnName -> {
-                        throw new IgniteInternalException(
-                                ErrorGroups.Index.INVALID_INDEX_DEFINITION_ERR, "Can't create table with duplicate columns: "
-                                + params.columns().stream().map(ColumnParams::name).collect(Collectors.joining(", "))
-                        );
-                    });
+            validateParams(params);
+
+            CatalogZoneDescriptor zone = getZone(catalog, Objects.requireNonNullElse(params.zone(), DEFAULT_ZONE_NAME));
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:273]: This and similar refactorings could have been done in another PR, they have nothing to do with the fix itself.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 3]
```diff
@@ -277,82 +259,62 @@ private Catalog catalogAt(long timestamp) {
...
-                throw new TableNotFoundException(schemaName, params.tableName());
-            }
+            CatalogTableDescriptor table = getTable(schema, params.tableName());
 
             List<UpdateEntry> updateEntries = new ArrayList<>();
 
             Arrays.stream(schema.indexes())
                     .filter(index -> index.tableId() == table.id())
-                    .forEach(index -> updateEntries.add(new DropIndexEntry(index.id())));
+                    .forEach(index -> updateEntries.add(new DropIndexEntry(index.id(), index.tableId())));
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:300]: Why do you need a tableId here?

**Review by AMashenkov — APPROVED** [review]

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractIndexCommandParams.java [hunk 4]
```diff
@@ -41,6 +47,20 @@ public String schemaName() {
...
+     * Gets table name.
+     */
+    public String tableName() {
+        return tableName;
+    }
+
+    /**
+     * Returns {@code true} if index is unique, {@code false} otherwise.
+     */
+    public boolean unique() {
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractIndexCommandParams.java:60]: What does this have to do with current fix?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java [hunk 5]
```diff
@@ -24,25 +24,27 @@
  * Add column event parameters contains descriptors of added columns.
  */
 public class AddColumnEventParameters extends CatalogEventParameters {
-
     private final int tableId;
-    private final List<CatalogTableColumnDescriptor> columnDescriptors;
+    private final List<CatalogTableColumnDescriptor> descriptors;
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java:28]: Why did you rename it? Old name was fine

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java [hunk 6]
```diff
@@ -23,32 +23,36 @@
  * Create table event parameters contains a column descriptor for the modified column.
  */
 public class AlterColumnEventParameters extends CatalogEventParameters {
-
     private final int tableId;
 
-    private final CatalogTableColumnDescriptor columnDescriptor;
+    private final CatalogTableColumnDescriptor descriptor;
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java:28]: And here. These are changes for the sake of changes

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java [hunk 7]
```diff
@@ -23,32 +23,36 @@
...
 
         this.tableId = tableId;
-        this.columnDescriptor = columnDescriptor;
+        this.descriptor = descriptor;
     }
 
-    /** Returns an id of a modified table. */
+    /**
+     * Returns ID of a modified table.
+     */
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java:47]: Was this necessary? Do we have code style guide that recommends this formatting?
If not, then you simply change someone-else's code by your own preference, without any other objective reason.
I wouldn't mind it, but this PR is 1000+ lines of changes, and some of them are completely unnecessary. This forces me to look at more code than I should.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterZoneEventParameters.java [hunk 8]
```diff
@@ -23,25 +23,25 @@
  * Alter zone event parameters contains a distribution zone descriptor for newly created distribution zone.
  */
 public class AlterZoneEventParameters extends CatalogEventParameters {
-
-    private final CatalogZoneDescriptor zoneDescriptor;
+    private final CatalogZoneDescriptor descriptor;
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterZoneEventParameters.java:26]: Same here. Why did you rename all these fields?

Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterZoneEventParameters.java:26]: I won't be leaving the same comment over and over again. I see these renames multiple times later.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/AlterZoneEntry.java [hunk 9]
```diff
@@ -37,12 +43,36 @@ public AlterZoneEntry(CatalogZoneDescriptor descriptor) {
...
+    }
+
+    @Override
+    public Catalog applyUpdate(Catalog catalog, VersionedUpdate update) {
+        return new Catalog(
+                update.version(),
+                update.activationTimestamp(),
+                catalog.objectIdGenState(),
+                catalog.zones().stream()
+                        .map(z -> z.id() == descriptor.id() ? descriptor : z)
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/AlterZoneEntry.java:70]: `z`, is this the original name?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog.storage;
+
+import org.apache.ignite.internal.catalog.events.CatalogEvent;
+import org.apache.ignite.internal.catalog.events.CatalogEventParameters;
+
+/**
+ * Interface for firing events.
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java:24]: for updates that require firing events, I guess

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java [hunk 11]
```diff
@@ -18,9 +18,11 @@
...
 
 import java.io.Serializable;
+import org.apache.ignite.internal.catalog.Catalog;
 
 /**
- * A marker interface describing a particular change within the {@link VersionedUpdate group}.
+ * Interface describing a particular change within the {@link VersionedUpdate group}.
  */
 public interface UpdateEntry extends Serializable {
+    Catalog applyUpdate(Catalog catalog, VersionedUpdate update);
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java:27]: Where's javadoc?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java [hunk 12]
```diff
@@ -167,7 +169,7 @@ private void restoreStateFromVault(OnUpdateHandler handler) {
 
             VersionedUpdate update = fromBytes(entry.value());
 
-            handler.handle(update);
+            handler.handle(update, metastore.appliedRevision());
```
Comment (ibessonov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java:172]: This has nothing to do with the fix. Please do it in a separate issue.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 13]
```diff
@@ -651,18 +456,12 @@ public CompletableFuture<Void> createDistributionZone(CreateZoneParams params) {
...
+            CatalogZoneDescriptor zone = getZone(catalog, params.zoneName());
 
-            CatalogZoneDescriptor zone = catalog.zone(zoneName);
-
-            if (zone == null) {
-                throw new DistributionZoneNotFoundException(zoneName);
-            }
-            if (zone.name().equals(CatalogService.DEFAULT_ZONE_NAME)) {
+            if (zone.name().equals(DEFAULT_ZONE_NAME)) {
                 //TODO IGNITE-19082 Can default zone be dropped?
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:465]: Fix it

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java (hunk 1 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java:1293]: Before dropping the table, I call `clearInvocations(eventListener);`

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractIndexCommandParams.java (hunk 4 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/AbstractIndexCommandParams.java:60]: Create ticket for this https://issues.apache.org/jira/browse/IGNITE-19799

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java (hunk 5 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java:28]: But not for me.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java (hunk 6 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java:28]: But not for me.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java (hunk 7 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java:47]: It is grammatically correct to use ID.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterZoneEventParameters.java (hunk 8 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterZoneEventParameters.java:26]: But not for me.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/AlterZoneEntry.java (hunk 9 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/AlterZoneEntry.java:70]: Yep

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java (hunk 10 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java:24]: fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java (hunk 12 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java:172]: But the catalog creates events that are descendants of `org.apache.ignite.internal.manager.EventParameters` that require `causalityToken`, the catalog now passes the correct value.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java (hunk 11 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java:27]: Trie to fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java (hunk 5 above)
Comment (korlov42) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java:28]: what is wrong with original name?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java [hunk 14]
```diff
@@ -24,25 +24,27 @@
...
 
         this.tableId = tableId;
-        this.columnDescriptors = columnDescriptors;
+        this.descriptors = descriptors;
     }
 
-    /** Returns table id. */
+    /**
+     * Returns table ID.
+     */
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AddColumnEventParameters.java:47]: single-line javadocs are perfectly fine, so, please, revert all such changes

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 15]
```diff
@@ -150,66 +138,61 @@ public CatalogServiceImpl(UpdateLog updateLog, HybridClock clock, long delayDura
         this.delayDurationMs = delayDurationMs;
     }
 
-    /** {@inheritDoc} */
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:153]: why did you remove this?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java [hunk 16]
```diff
@@ -36,7 +36,7 @@
  * <p>TBD: events
  */
 public interface CatalogService {
-    String PUBLIC = "PUBLIC";
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java:39]: was it really necessary to rename this constant?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 17]
```diff
@@ -277,29 +254,18 @@ private Catalog catalogAt(long timestamp) {
...
-                                ErrorGroups.Index.INVALID_INDEX_DEFINITION_ERR, "Can't create table with duplicate columns: "
-                                + params.columns().stream().map(ColumnParams::name).collect(Collectors.joining(", "))
-                        );
-                    });
+            validateParams(params);
 
-            String zoneName = Objects.requireNonNullElse(params.zone(), CatalogService.DEFAULT_ZONE_NAME);
-
-            CatalogZoneDescriptor zone = Objects.requireNonNull(catalog.zone(zoneName), "No zone found: " + zoneName);
+            CatalogZoneDescriptor zone = getZone(catalog, Objects.requireNonNullElse(params.zone(), DEFAULT_ZONE_NAME));
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:268]: All this refactoring has nothing to do with problem to be solved within the ticket (IGNITE-19641). So, please, move it to another patch

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java [hunk 18]
```diff
@@ -1089,4 +644,195 @@ private static void throwUnsupportedDdl(String msg, Object... params) {
     interface UpdateProducer {
         List<UpdateEntry> get(Catalog catalog);
     }
+
+    private static CatalogSchemaDescriptor getSchema(Catalog catalog, @Nullable String schemaName) {
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java:648]: all this new lines have nothing to do with the problem described in ticket, so, please, move it to another patch

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/CatalogEventParameters.java [hunk 19]
```diff
@@ -23,12 +23,24 @@
...
+    private final int catalogVersion;
+
     /**
      * Constructor.
      *
      * @param causalityToken Causality token.
+     * @param catalogVersion Catalog version.
      */
-    public CatalogEventParameters(long causalityToken) {
+    public CatalogEventParameters(long causalityToken, int catalogVersion) {
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/CatalogEventParameters.java:34]: why version of the catalog can't be causality token?

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java (hunk 1 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogServiceSelfTest.java]: Revert it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.apache.ignite.internal.catalog.storage;
+
+import org.apache.ignite.internal.catalog.events.CatalogEvent;
+import org.apache.ignite.internal.catalog.events.CatalogEventParameters;
+
+/**
+ * Interface for updates that require firing events.
+ */
+public interface CatalogFireEvent {
```
Comment (korlov42) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/CatalogFireEvent.java:26]: Within the package 'storage' there is no single class whose name has prefix "Catalog", thus let's make naming consistent.
Also, `fireEvent` is a good name for method to fire event, but not for the object that provides the event. Have you considered the names like `Fireable`, `EventSource`, or `EventProvider`?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/CatalogEventParameters.java (hunk 19 above)
Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/CatalogEventParameters.java:34]: I was involved in it. Right now we only have meta-storage revision as a causality token. If we use two different entities, calling it the same name, it will lead to confusion. It's easier and safer to migrate meta-storage revision into a catalog version later with a single commit

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java (hunk 2 above)
Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java]: Refactoring that has nothing to do with the fix you provide. `createTable` is not related to anything described in the original issue, just as any other "alter" methods. You only changed the part where updates are being converted to events. I'm all for refactoring that part, it's good, but everything else is just confusing and unnecessary

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java (hunk 7 above)
Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/events/AlterColumnEventParameters.java:47]: Ok, but converting
```
/** Foo. */
```
into
```
/**
 * Foo.
 */
```
is not related to grammar

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java (hunk 12 above)
Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateLogImpl.java:172]: It's only required now because you added it, but ok, this change is not as big as the rest

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java (hunk 16 above)
Comment (ibessonov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogService.java:39]: I would say that this remove is justified, but in another PR maybe. Name "PUBLIC" says nothing about what it is, and there's no javadoc to clarify things either.

**Review by ibessonov — APPROVED** [review]

**Review by korlov42 — COMMENTED** [review]
we definitely need regression test for this

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java [hunk 21]
```diff
@@ -18,9 +18,18 @@
...
  */
 public interface UpdateEntry extends Serializable {
+    /**
+     * Applies own change to the catalog.
+     *
+     * @param catalog Current catalog.
+     * @param update Group of changes that relates to specified version.
+     * @return New catalog.
+     */
+    Catalog applyUpdate(Catalog catalog, VersionedUpdate update);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/UpdateEntry.java:34]: Fix it

**Review by korlov42 — APPROVED** [review]
